source/parser.hera

```diff
@@ -6284,37 +6296,53 @@ Init
     }
 
     function expressionizeIteration(exp) {
-      const resultsRef = {
-        type: "Ref",
-        base: "results",
-        id: "results",
-      }
+      const i = exp.children.indexOf(exp.block)
+      if (exp.subtype === "DoStatement") {
+        // Just wrap with IIFE
+        insertReturn(exp.block)
+        exp.children.splice(i, 1, "(()=>", ...exp.children, ")()")
+      } else {
+        const resultsRef = {
+          type: "Ref",
+          base: "results",
+          id: "results",
+        }
+
+        // insert `results.push` to gather results array
+        insertPush(exp.block, resultsRef)
 
-      // insert `results.push` to gather results array
-      insertPush(exp.block, resultsRef)
-      // Wrap with IIFE
-      exp.children = ["(()=>{const ", resultsRef, "=[];", ...exp.children, "; return ", resultsRef, "})()"]
+        // Wrap with IIFE
+        exp.children.splice(i, 1, "(()=>{const ", resultsRef, "=[];", ...exp.children, "; return ", resultsRef, "})()")
+      }
     }
 
     function wrapIterationReturningResults(statement, outerRef) {
-      const resultsRef = {
-        type: "Ref",
-        base: "results",
-        id: "results",
-      }
+      if (statement.type === "DoStatement") {
+        if (outerRef) {
+          insertPush(statement.block, outerRef)
+        } else {
+          insertReturn(statement.block)
+        }
+      } else {
```

Comment:
We can avoid `else` and nesting further with an early return at the end of the `if`.